docs/src/main/asciidoc/extension-authors-guide.adoc

```diff
@@ -296,40 +296,141 @@ You will also need to configure the `maven-compiler-plugin` to detect the `quark
 
 === Build Step Processors
 
-Work is done at deployment time by producing and consuming instances of `io.quarkus.builder.item.BuildItem`. This is done
-by creating a class that has method(s) annotated with `io.quarkus.deployment.annotations.BuildStep`. These classes can
-consume items by injection, and produce items by either returning them from the method or by injecting an
-instance of `io.quarkus.deployment.annotations.BuildProducer` for the produced type. These processors can also record
-bytecode invocations, which is mapped to a `BuildItem` transparently.
+Work is done at augmentation time by _build steps_ which produce and consume _build items_.  The build steps found in
+the deployment modules that correspond to the extensions in the project build are automatically wired together and executed
+to produce the final build artifact(s).
 
-There are two distinct types of `BuildItem`, `SimpleBuildItem` and `MultiBuildItem`. `SimpleBuildItem` will only ever
-have a single instance created, while `MultiBuildItem` can have many instances.
+==== Build steps
```

Comment:
Thanks, the doc is a massive step up!

However we need have some concrete examples so that people can read the concrete example and then read the abstract documentation and not be totally lost.
One simple, one multi,  one virtual build step and one @Weak usage based on real live example would go a long way.
I'd recommend to put these examples before and the detailed explanations after.

PS: I understand that you started not from the best guide and therefore more work falls on you by side effect. Sorry about that :)

Reply:
It's no problem, I'll come up with something. :)